### PR TITLE
Adding JWT Security

### PR DESCRIPTION
**Pull request by drilonsmajli**

### CODE REVIEW

File: src/main/java/ProjektiPizza/demo/security/JwtSuccessHandler.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+import org.springframework.security.core.Authentication;
+import org.springframework.security.web.authentication.AuthenticationSuccessHandler;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class JwtSuccessHandler implements AuthenticationSuccessHandler {
+    @Override
+    public void onAuthenticationSuccess(HttpServletRequest httpServletRequest, HttpServletResponse httpServletResponse, Authentication authentication) {
+        System.out.println("Successfully Authentication");
```
Comment (ermalism-zz) [inline src/main/java/ProjektiPizza/demo/security/JwtSuccessHandler.java]: Can we use a logger here instead of printing with the system.out?

File: src/main/java/ProjektiPizza/demo/security/JwtValidator.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+                    .parseClaimsJws(token)
+                    .getBody();
+
+            user = new User();
+
+            user.setUsername(body.getSubject());
+            user.setFullName(body.getSubject());
+            user.setPassword(body.getSubject());
+        } catch (Exception e) {
+            System.out.println(e);
```
Comment (ermalism-zz) [inline src/main/java/ProjektiPizza/demo/security/JwtValidator.java]: Also here!

File: src/main/java/ProjektiPizza/demo/security/JwtValidator.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import ProjektiPizza.demo.entity.User;
+import io.jsonwebtoken.Claims;
+import io.jsonwebtoken.Jwts;
+import org.springframework.stereotype.Component;
+
+@Component
+public class JwtValidator {
+
+
+    private String secret = "youtube";
```
Comment (ermalism-zz) [inline src/main/java/ProjektiPizza/demo/security/JwtValidator.java]: Maybe it would be better if we save these in a file with constants?

File: src/main/java/ProjektiPizza/demo/security/JwtGenerator.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+public class JwtGenerator {
+
+    public String generate(User user) {
+        Claims claims = Jwts.claims()
+                .setSubject(user.getUsername());
+        claims.setSubject(user.getPassword());
+        claims.setSubject(user.getFullName());
+        return Jwts.builder()
+                .setClaims(claims)
+                .signWith(SignatureAlgorithm.HS512, "youtube")
```
Comment (ermalism-zz) [inline src/main/java/ProjektiPizza/demo/security/JwtGenerator.java]: That constants class could be used here for example without duplicating the string "youtube" here

File: src/main/java/ProjektiPizza/demo/security/JwtAuthenticationTokenFilter.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+public class JwtAuthenticationTokenFilter extends AbstractAuthenticationProcessingFilter {
+
+    public JwtAuthenticationTokenFilter() {
+        super("/api/**");
```
Comment (ermalism-zz) [inline src/main/java/ProjektiPizza/demo/security/JwtAuthenticationTokenFilter.java:17]: Is the authentication required if I try to hit `/test-api/`?

File: src/main/java/ProjektiPizza/demo/security/JwtAuthenticationProvider.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+    @Override
+    protected User retrieveUser(String username, UsernamePasswordAuthenticationToken usernamePasswordAuthenticationToken) throws AuthenticationException {
+        JwtAuthenticationToken jwtAuthenticationToken = (JwtAuthenticationToken) usernamePasswordAuthenticationToken;
+        String token = jwtAuthenticationToken.getToken();
+
+        User user = validator.validate(token);
+
+        if (user == null) {
+            throw new RuntimeException("JWT token is incorrect");
```
Comment (ermalism-zz) [inline src/main/java/ProjektiPizza/demo/security/JwtAuthenticationProvider.java:31]: This returns the response to the client as string, right? Maybe it would be better if we do this as a constructed JSON.

File: src/main/java/ProjektiPizza/demo/Security/SecurityConfiguration.java [hunk 7]
```diff
@@ -1,48 +0,0 @@
-package ProjektiPizza.demo.Security;
```
Comment (ermalism-zz) [inline src/main/java/ProjektiPizza/demo/Security/SecurityConfiguration.java:1]: Why is this deleted? What changed in the code that you could afford deleting this? Can I hit it from the client that is in another IP (is CORS enabled)?

File: src/main/java/ProjektiPizza/demo/security/JwtSuccessHandler.java (hunk 1 above)
Comment (drilonsmajli) [reply src/main/java/ProjektiPizza/demo/security/JwtSuccessHandler.java]: Yes we can.

File: src/main/java/ProjektiPizza/demo/security/JwtValidator.java (hunk 2 above)
Comment (drilonsmajli) [reply src/main/java/ProjektiPizza/demo/security/JwtValidator.java]: yes. it will print in console.

File: src/main/java/ProjektiPizza/demo/Security/SecurityConfiguration.java (hunk 7 above)
Comment (drilonsmajli) [reply src/main/java/ProjektiPizza/demo/Security/SecurityConfiguration.java:1]: we change only first letter Security to security

File: src/main/java/ProjektiPizza/demo/security/JwtAuthenticationTokenFilter.java (hunk 5 above)
Comment (drilonsmajli) [reply src/main/java/ProjektiPizza/demo/security/JwtAuthenticationTokenFilter.java:17]: no it will't work.
it works only for request api/another request

File: src/main/java/ProjektiPizza/demo/security/JwtAuthenticationProvider.java (hunk 6 above)
Comment (drilonsmajli) [reply src/main/java/ProjektiPizza/demo/security/JwtAuthenticationProvider.java:31]: i'll try to constructed JSON

File: src/main/java/ProjektiPizza/demo/security/JwtGenerator.java (hunk 4 above)
Comment (drilonsmajli) [reply src/main/java/ProjektiPizza/demo/security/JwtGenerator.java]: you're right.
